### PR TITLE
Implemented create survey feature

### PR DESCRIPTION
**Pull request by madeye-matt**
- can be found at /surveys/create (no link from survey page)

This PR is feature complete but I need to add tests (once I've learned how ...)

### CODE REVIEW

**Comment by madeye-matt** [comment]
[US102: INT User to Create a CE - After R17/18 (M)](https://trello.com/c/9VgewZXM/77-us102-int-user-to-create-a-ce-after-r17-18-m)

**Comment by madeye-matt** [comment]
[US101: INT User to Create a Survey - After R17/18 (L)](https://trello.com/c/p7lh6cuO/68-us101-int-user-to-create-a-survey-after-r17-18-l)

File: response_operations_ui/controllers/survey_controllers.py [hunk 1]
```diff
@@ -100,3 +100,39 @@ def update_survey_details(survey_ref, short_name, long_name):
         raise ApiError(response)
 
     logger.debug('Successfully updated survey details', survey_ref=survey_ref)
+
+
+def get_legal_basis_list():
+    logger.debug('Retrieving legal basis list')
+    url = f'{app.config["SURVEY_SERVICE_URL"]}/legal-bases'
+    response = requests.get(url, auth=(app.config['SURVEY_SERVICE_USERNAME'], app.config['SURVEY_SERVICE_PASSWORD']))
```
Comment (gemmai95) [inline response_operations_ui/controllers/survey_controllers.py]: Will this not always be the same as the basic auth username and password?

Comment (madeye-matt) [reply response_operations_ui/controllers/survey_controllers.py]: No, I asked Dave about this and he said each service would have it's own username password

File: response_operations_ui/views/surveys.py [hunk 2]
```diff
@@ -84,4 +96,45 @@ def edit_survey_details(short_name):
         survey_controllers.update_survey_details(form.get('hidden_survey_ref'),
                                                  form.get('short_name'),
                                                  form.get('long_name'))
-        return redirect(url_for('surveys_bp.view_surveys', short_name=short_name, survey_changed='True'))
+        return redirect(url_for('surveys_bp.view_surveys', short_name=short_name, message_key='survey_changed'))
+
+
+@surveys_bp.route('/create', methods=['GET'])
+@login_required
+def show_create_survey():
```
[the remark this answers is not in the document]
Comment (madeye-matt) [reply response_operations_ui/views/surveys.py]: Hmmm I used the same pattern as the edit-survey-details code - does the same observation apply to that or am I missing something here?

[the remark this answers is not in the document]
Comment (madeye-matt) [reply response_operations_ui/views/surveys.py]: Ok this one is done

File: response_operations_ui/controllers/survey_controllers.py [hunk 3]
```diff
@@ -100,3 +100,39 @@ def update_survey_details(survey_ref, short_name, long_name):
...
+    if response.status_code != 200:
+        raise ApiError(response)
+
+    lbs = [(lb['ref'], lb['longName']) for lb in response.json()]
+    logger.debug('Successfully retrieved surveys list: {}'.format(lbs))
```
Comment (rhiGriff) [inline response_operations_ui/controllers/survey_controllers.py]: Log lines should be in key value pairs.

File: response_operations_ui/error_handlers.py [hunk 4]
```diff
@@ -12,7 +12,7 @@
 
 @app.errorhandler(ApiError)
 def api_error(error):
-    logger.error('Api failed to retrieve required data', url=error.url,
+    logger.error('Api failed to retrieve required data - {}'.format(error.message), url=error.url,
```
Comment (rhiGriff) [inline response_operations_ui/error_handlers.py]: Shouldn't be logging a url

File: response_operations_ui/views/surveys.py [hunk 5]
```diff
@@ -84,4 +96,45 @@ def edit_survey_details(short_name):
...
+def create_survey():
+    form = CreateSurveyDetailsForm(form=request.form)
+    if not form.validate():
+        return render_template('create-survey.html', form=form, errors=form.errors.items(),
+                               survey_ref=request.form.get('survey_ref'),
+                               long_name=request.form.get('long_name'),
+                               short_name=request.form.get('short_name'),
+                               legal_basis=request.form.get('legal_basis'))
+    else:
+        logger.debug('create-survey form: {}'.format(form))
```
Comment (rhiGriff) [inline response_operations_ui/views/surveys.py]: Same again this should be key value pairs

**Comment by gemmai95** [comment]
Can you add these config changes to the confluence doc for new config? https://digitaleq.atlassian.net/wiki/spaces/RASB/pages/406257665/Pending+Release+Config+Changes

File: response_operations_ui/controllers/survey_controllers.py [hunk 6]
```diff
@@ -100,3 +100,40 @@ def update_survey_details(survey_ref, short_name, long_name):
...
+        "legalBasisRef": legal_basis
+    }
+
+    response = requests.post(
+        url,
+        json=survey_details,
+        auth=(app.config['SURVEY_SERVICE_USERNAME'], app.config['SURVEY_SERVICE_PASSWORD']))
+
+    if response.status_code != 201:
+        logger.debug("Raising ApiError for response code {}".format(response.status_code))
```
Comment (rhiGriff) [inline response_operations_ui/controllers/survey_controllers.py]: This log line should also be a key value pair eg, status=status_code

File: config.py [hunk 7]
```diff
@@ -10,6 +10,9 @@ class Config(object):
     RESPONSE_OPERATIONS_UI_SECRET = os.getenv('RESPONSE_OPERATIONS_UI_SECRET', "secret")
     SECURE_MESSAGE_URL = os.getenv('SECURE_MESSAGE_URL')
     BACKSTAGE_API_URL = os.getenv('BACKSTAGE_API_URL')
+    SURVEY_SERVICE_URL = os.getenv('SURVEY_SERVICE_URL')
+    SURVEY_SERVICE_USERNAME = os.getenv('SURVEY_SERVICE_USERNAME')
+    SURVEY_SERVICE_PASSWORD = os.getenv('SURVEY_SERVICE_PASSWORD')
```
Comment (gemmai95) [inline config.py]: There has now been a decision to remove service from config names

File: response_operations_ui/controllers/survey_controllers.py [hunk 8]
```diff
@@ -100,3 +100,40 @@ def update_survey_details(survey_ref, short_name, long_name):
         raise ApiError(response)
 
     logger.debug('Successfully updated survey details', survey_ref=survey_ref)
+
+
+def get_legal_basis_list():
+    logger.debug('Retrieving legal basis list')
+    url = f'{app.config["SURVEY_URL"]}/legal-bases'
+    response = requests.get(url, auth=(app.config['SURVEY_USERNAME'], app.config['SURVEY_PASSWORD']))
```
Comment (tej99) [inline response_operations_ui/controllers/survey_controllers.py]: Might be better to put SURVEY_BASIC_AUTH = (SURVEY_USERNAME, SURVEY_PASSWORD) in the config.

Comment (madeye-matt) [reply response_operations_ui/controllers/survey_controllers.py]: Tis done (I didn't know you could do that ... :-) )

**Pull request merged**